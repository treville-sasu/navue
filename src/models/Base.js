export class Model {
  // eslint-disable-next-line no-unused-vars
  constructor({ type, ...properties } = {}) {
    Object.assign(this, properties);
  }

  toJSON() {
    return {
      ...this,
      type: this.constructor.name
    };
  }

  // TODO write specs
  clone() {
    return this.constructor.from(JSON.parse(JSON.stringify(this)));
  }

  static from({ type, ...properties } = {}) {
    if (arguments[0] instanceof this) return arguments[0];
    else if (type == this.name) return new this(properties);
    else throw `Invalid data : 'type' should be '${this.name}' got '${type}'`;
  }
}

// TODO : implementation for when Vuejs Suppots Array-like objects.
// export class Store extends Array {
export class Store {
  constructor(properties = {}, ...items) {
    this.items = items;
    Object.assign(this, properties);
  }

  // get items() {
  //   return [...this];
  // }

  [Symbol.iterator]() {
    return this.items[Symbol.iterator]();
  }

  get length() {
    return this.items.length;
  }

  valueOf() {
    return this.items;
  }

  get properties() {
    // eslint-disable-next-line no-unused-vars
    let { items, ...props } = { ...this };
    return props;
    // return Object.fromEntries(
    //   Object.entries(this).filter(([key]) => isNaN(key))
    // );
  }

  get first() {
    return this.items[0];
  }

  get last() {
    return this.items[this.length - 1];
  }

  add(item, index) {
    index ? this.items.splice(index, 0, item) : this.items.push(item);
    return this.length;
  }

  remove(item, index) {
    if (index === undefined) index = this.items.indexOf(item);
    return index >= 0 ? this.items.splice(index, 1).length == 1 : false;
  }

  keep(filter) {
    let i = this.items.length;
    while (i--) {
      let result = filter.call(this.items, this.items[i]);

      if (result == undefined) this.items.splice(i, 1);
      else if (!Object.is(result, this.items[i]))
        this.items.splice(i, 1, result);
    }
    return this;
  }

  toJSON() {
    return {
      ...this.properties,
      items: this.items,
      type: this.constructor.name
    };
  }

  static from({ type, items = [], ...properties }, ...constructors) {
    if (type != this.name)
      throw `Invalid data : 'type' should be '${this.name}' got '${type}'`;
    else {
      let constructor = constructors.shift();
      return new this(
        properties,
        ...items.map(i => {
          return constructor ? constructor.from(i, ...constructors) : i;
        })
      );
    }
  }
}
