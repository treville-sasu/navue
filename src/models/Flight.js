import { Journey } from "@/models/Journey";
import { Location } from "@/models/Location";
import { Branch } from "./Journey";

export class Flight extends Journey {
  constructor(
    { name, notes, checked = [], milestones = [] } = {},
    ...locations
  ) {
    // let branches;
    // if (locations.every(a => a instanceof Location))
    //   branches = [new Branch(Location, {}, ...locations)];
    // else if (
    //   locations.every(
    //     a => a instanceof Array && a.every(b => b instanceof Branch)
    //   )
    // )
    //   branches = locations.flat();
    let branches = locations;
    super({ name, notes, checked, milestones }, ...branches);
  }

  addBranch(locations, properties) {
    this.branches.push(new Branch(Location, properties, ...locations));
    return this;
  }

  get duration() {
    return (
      this.branches.reduce((total, b) => {
        return (total +=
          b.last().properties.timestamp - b.first().properties.timestamp);
      }, 0) / 1000
    );
  }

  // static from({ type, locations, features, ...properties } = {}) {
  //   if (arguments[0] instanceof this) return arguments[0];

  //   switch (type) {
  //     case this.name:
  //       if (locations)
  //         locations = locations.map(t => t.map(l => Location.from(l)));

  //       break;
  //     case "FeatureCollection":
  //       locations = [features.map(l => Location.from(l))];
  //       // properties = { ...properties };
  //       break;

  //     default:
  //       throw `Invalid data : 'type' should be 'FeatureCollection' or '${this.name}' got '${type}'`;
  //   }

  //   return new this(properties, ...locations);
  // }
}
