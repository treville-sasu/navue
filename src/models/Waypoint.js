import { Model } from "@/models/Base.js";
import { Altitude, Distance, Azimuth } from "@/models/Quantities.js";

import { point, round } from "@turf/helpers";
import rhumbDistance from "@turf/rhumb-distance";
import rhumbBearing from "@turf/rhumb-bearing";
import rhumbDestination from "@turf/rhumb-destination";
import truncate from "@turf/truncate";
import greatCircle from "@turf/great-circle";

export class Waypoint extends Model {
  constructor([longitude, latitude, altitude] = [], properties = {}) {
    const coordinates = [round(longitude, 6), round(latitude, 6)];

    if (altitude !== undefined && !properties["altitude"]) {
      properties["altitude"] = new Altitude(altitude, "m", {
        reference: "WGS84"
      });
    } else if (properties["altitude"]) {
      properties["altitude"] = Altitude.from(properties["altitude"]);
      altitude = round(Number(altitude));
    }

    if (altitude) coordinates.push(round(Number(altitude)));
    super(point(coordinates, properties));
  }

  set longitude(val) {
    this.geometry.coordinates[0] = round(val, 6);
  }

  set latitude(val) {
    this.geometry.coordinates[1] = round(val, 6);
  }

  get altitude() {
    return this.geometry.coordinates[2];
  }

  set altitude(val) {
    this.geometry.coordinates[2] = round(val);
    this.properties.altitude = val;
  }

  get lngLat() {
    return {
      lng: this.geometry.coordinates[0],
      lat: this.geometry.coordinates[1]
    };
  }

  set lngLat({ lng, lat }) {
    this.longitude = lng;
    this.latitude = lat;
  }

  distanceTo(wp) {
    return new Distance(
      round(
        rhumbDistance(this.geometry.coordinates, wp.geometry.coordinates, {
          units: "meters"
        })
      )
    );
  }

  bearingTo(wp) {
    return new Azimuth(
      rhumbBearing(this.geometry.coordinates, wp.geometry.coordinates),
      "°"
    );
  }

  greatCircleTo(wp, properties) {
    let gc = truncate(
      greatCircle(this.geometry.coordinates, wp.geometry.coordinates, {
        npoints: 8,
        properties
      })
    );

    gc.properties = {
      ...gc.properties,
      distance: this.distanceTo(wp),
      bearing: this.bearingTo(wp)
    };

    return gc;
  }

  destinationPoint(distance, heading, properties) {
    const coords = rhumbDestination(
      this.geometry.coordinates,
      new Distance(distance),
      new Azimuth(heading).to("°"),
      {
        units: "meters"
      }
    ).geometry.coordinates;
    if (this.geometry.coordinates[2]) coords[2] = this.geometry.coordinates[2];
    return new this.constructor(coords, properties);
  }

  toGeoJSON(type) {
    switch (type) {
      case "Position":
        return this.geometry.coordinates;
      case "Point":
        return this.geometry;
      default:
        return {
          ...this,
          type: "Feature"
        };
    }
  }

  static from({
    type,
    geometry,
    coordinates,
    properties = {},
    ...otherProps
  } = {}) {
    if (arguments[0] instanceof this) return arguments[0];

    switch (type) {
      case this.name:
      case "Feature":
        if (geometry && geometry.type == "Point")
          coordinates = geometry.coordinates;
        else
          throw `Invalid data : 'geometry.type' should be 'Point' got '${geometry &&
          geometry.type}'`;
        break;
      case "Point":
        // coordinates = coordinates;
        break;
      default:
        throw `Invalid data : 'type' should be 'Feature', 'Point' or '${this.name}' got '${type}'`;
    }
    return new this(coordinates, {
      ...otherProps,
      ...properties
    });
  }

  static fromEvent({ lngLat: { lng, lat } } = {}) {
    return new this([lng, lat]);
  }
}
