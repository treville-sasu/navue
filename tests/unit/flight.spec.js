import { Flight } from "@/models/Flight.js";
import { Location } from "@/models/Waypoint.js";
import { Model } from "../../src/models/Base";

import theFlight from "../flight.json";

describe("flight", () => {
  it("extends Model", () => {
    expect(Flight.prototype).toBeInstanceOf(Model);
  });
  it("create with defaults", () => {
    expect(new Flight()).toHaveProperty("name", undefined);
    expect(new Flight()).toHaveProperty("notes", undefined);
    expect(new Flight()).toHaveProperty("checked", expect.any(Array));
    expect(new Flight()).toHaveProperty(
      "locations",
      expect.arrayContaining([])
    );
    expect(new Flight()).toHaveProperty("milestones", expect.any(Array));
  });
  it("create with locations & without properties", () => {
    expect(
      new Flight(
        {},
        new Location(),
        new Location(),
        new Location(),
        new Location()
      )
    ).toHaveProperty("locations", [
      [new Location(), new Location(), new Location(), new Location()]
    ]);
    expect(
      new Flight({}, [
        new Location(),
        new Location(),
        new Location(),
        new Location()
      ])
    ).toHaveProperty("locations", [
      [new Location(), new Location(), new Location(), new Location()]
    ]);
    expect(
      new Flight(
        {},
        [new Location(), new Location()],
        [new Location(), new Location()]
      )
    ).toHaveProperty("locations", [
      [new Location(), new Location()],
      [new Location(), new Location()]
    ]);
  });

  it("create with properties & rejects unvalid property", () => {
    const flight = new Flight({
      name: "test flight",
      notes: "model",
      manufacturer: "some text", //not allowed
      checked: [],
      milestones: []
    });
    expect(flight).toHaveProperty("name", "test flight");
    expect(flight).toHaveProperty("notes", "model");
    expect(flight).toHaveProperty("checked", []);
    expect(flight).toHaveProperty("milestones", []);
    expect(flight).toHaveProperty("locations");

    expect(flight).not.toHaveProperty("manufacturer", "some text");
  });

  it("addLocation", () => {
    expect(
      new Flight({}, new Location(0, 0)).addLocation(new Location(1, 1))
    ).toHaveProperty("locations", [[new Location(0, 0), new Location(1, 1)]]);
  });

  it("addTrace", () => {
    expect(
      new Flight().addTrace(new Location(0, 0), new Location(1, 1))
    ).toHaveProperty("locations", [[new Location(0, 0), new Location(1, 1)]]);
    expect(
      new Flight().addTrace([new Location(0, 0), new Location(1, 1)])
    ).toHaveProperty("locations", [[new Location(0, 0), new Location(1, 1)]]);
  });

  it("return trace as a multiArray of latitude, longitude, altitude", () => {
    expect(
      new Flight(
        {},
        [new Location(-1, 0, 0), new Location(0, 1, 1)],
        [new Location(0, -1, 1), new Location(1, 1, 0)]
      ).geom
    ).toStrictEqual([
      [
        [-1, 0, 0],
        [0, 1, 1]
      ],
      [
        [0, -1, 1],
        [1, 1, 0]
      ]
    ]);
  });

  it("return a GeoJSON object", () => {
    let testFlight = new Flight({ name: "test flight", notes: "short one" }, [
      [new Location(-1, 0, 0), new Location(0, 1, 1)],
      [new Location(0, -1, 1), new Location(1, 1, 0)]
    ]);
    expect(testFlight.toGeoJSON("MultiPoint")).toStrictEqual({
      type: "MultiPoint",
      coordinates: [
        [-1, 0, 0],
        [0, 1, 1],
        [0, -1, 1],
        [1, 1, 0]
      ]
    });
    expect(testFlight.toGeoJSON("MultiLineString")).toStrictEqual({
      type: "MultiLineString",
      coordinates: [
        [
          [-1, 0, 0],
          [0, 1, 1]
        ],
        [
          [0, -1, 1],
          [1, 1, 0]
        ]
      ]
    });
    expect(testFlight.toGeoJSON("Feature")).toStrictEqual({
      type: "Feature",
      properties: {
        checked: [],
        milestones: [],
        name: "test flight",
        notes: "short one"
      },
      geometry: {
        type: "MultiLineString",
        coordinates: [
          [
            [-1, 0, 0],
            [0, 1, 1]
          ],
          [
            [0, -1, 1],
            [1, 1, 0]
          ]
        ]
      }
    });
    expect(testFlight.toGeoJSON()).toStrictEqual(
      testFlight.toGeoJSON("Feature")
    );
  });

  it("calculate bounds of trace", () => {
    expect(
      new Flight({ name: "test flight", notes: "short one" }, [
        [new Location(-1, 0, 0), new Location(0, 1, 1)],
        [new Location(0, -1, 1), new Location(1, 1, 0)]
      ]).toBounds()
    ).toStrictEqual([
      [-1, -1],
      [1, 1]
    ]);
  });

  it.todo("duration");

  it("create a Flight from literal", () => {
    let anyFlight = new Flight();

    expect(() => Flight.from({})).toThrow(
      "Invalid data : 'type' should be 'Flight' got 'undefined'"
    );

    expect(Flight.from(anyFlight)).toBe(anyFlight);
    expect(Flight.from(theFlight)).toStrictEqual(theFlight);
  });
});
