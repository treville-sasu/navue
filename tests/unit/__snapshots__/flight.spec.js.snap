// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flight create a Flight from literal 1`] = `
Object {
  "_id": "Flight-1621437190027",
  "_rev": "1-415c19c5b5d1a534dfc429735749ba46",
  "checked": Array [],
  "locations": Array [
    Array [
      Object {
        "accuracy": Object {
          "precision": 16,
          "type": "Distance",
          "unit": "m",
          "value": 74.88337485607397,
        },
        "altitude": Object {
          "accuracy": 75.47549119035168,
          "precision": 15,
          "reference": "WGS84",
          "type": "Altitude",
          "unit": "m",
          "value": 974.745483588706,
        },
        "heading": Object {
          "precision": 17,
          "type": "Azimuth",
          "unit": "°",
          "value": 322.91307315742955,
        },
        "latitude": 43.42173780154349,
        "longitude": -0.37592911818056424,
        "speed": Object {
          "precision": 16,
          "type": "Speed",
          "unit": "m/s",
          "value": 49.77274617746323,
        },
        "timestamp": 1621436827165,
        "type": "Location",
      },
      Object {
        "accuracy": Object {
          "precision": 15,
          "type": "Distance",
          "unit": "m",
          "value": 74.5419678997506,
        },
        "altitude": Object {
          "accuracy": 75.48275531774489,
          "precision": 17,
          "reference": "WGS84",
          "type": "Altitude",
          "unit": "m",
          "value": 1014.2498616293805,
        },
        "heading": Object {
          "precision": 16,
          "type": "Azimuth",
          "unit": "°",
          "value": 324.3007601189933,
        },
        "latitude": 43.35158218238678,
        "longitude": -0.394352740677683,
        "speed": Object {
          "precision": 16,
          "type": "Speed",
          "unit": "m/s",
          "value": 50.19929914311654,
        },
        "timestamp": 1621436832083,
        "type": "Location",
      },
    ],
    Array [
      Object {
        "accuracy": Object {
          "precision": 16,
          "type": "Distance",
          "unit": "m",
          "value": 74.55342572630838,
        },
        "altitude": Object {
          "accuracy": 74.73206670357781,
          "precision": 16,
          "reference": "WGS84",
          "type": "Altitude",
          "unit": "m",
          "value": 1030.762731215022,
        },
        "heading": Object {
          "precision": 15,
          "type": "Azimuth",
          "unit": "°",
          "value": 89.0530377229343,
        },
        "latitude": 43.359593020493556,
        "longitude": -0.3841041381961341,
        "speed": Object {
          "precision": 16,
          "type": "Speed",
          "unit": "m/s",
          "value": 49.65466589164299,
        },
        "timestamp": 1621436897234,
        "type": "Location",
      },
    ],
  ],
  "milestones": Array [],
  "name": "Fixture",
  "notes": "some stuff",
  "type": "Flight",
}
`;
